### ISSUE TITLE
webtc2 too many results

### ISSUE THREAD

**Issue by funderburkjim** [sanskrit-lexicon/Cologne#286]
The webmaster pointed out an issue with advanced search display.

To summarize,  he mentioned that there are occasional times when abnormal cpu usage occurs, in conjunction with very long running requests.

He provided an example user url which caused such an event.  It involved the Advanced Search.

After some examination, two changes were made in advanced search that (a) fix the specific example query, and (b) provide a limitation that should limit cpu usage for other advanced search queries that the specific fix doesn't handle.  Details described in further comments.

**Comment by funderburkjim** [comment]
## Description of problem
Here is a screenshot of the problem, on local XAMPP installation, 

![image](https://user-images.githubusercontent.com/6393033/65364818-2452aa80-dbe2-11e9-81fc-0546de61d866.png)

* The user input a headword in Devanagari -- you can understand how this might happen, but
   in this display Devanagari is not accepted as input, and, even if it were,  the proper interpretation
   of the Devanagari input would, with current techniques, require that the input type be set 
   at 'Deva', not Harvard-Kyoto.
* The user has chosen 'all' for the maximum number of matches.
*  The Xampp server being used for the test is configured to allow a maximum of 30 seconds for
    a given execution time.   This accounts for the error message.
    *  How this would be handled on the Cologne server would depend on how that server is
       configured.

**Comment by funderburkjim** [comment]
## Why the example code runs so long.

When the example search is initiated,  a regular expression is constructed based on the various user choices of settings.  Among other things,  non-ascii characters of the input are removed; in our case the input is ऐश्वर्य  , which has only non-ascii characters,  so removing them leaves only the empty string.   The result is that the regexp is `[^a-zA-Z0-9]()[^a-zA-Z0-9]` .  This regexp is used to search every line of query_dump.  And every line in query_dump having (for example) two spaces would match.  Result is that almost every line matches!   

Then, since the user is requesting *all* matches,  the number of matching headwords will be almost every headword in the dictionary -- e.g. for mw 200,000 or so.   
Finally the program will generate html for all these headwords (probably several hundred megabytes of html) will attempt to be sent to the user's browser.

**Comment by funderburkjim** [comment]
## First fix

The first fix simply checks if, after removing non-ascii characters, the user input is the empty string.
If so, the program immediately fails.

Incidentally, the reason for removing characters from the user's input is to attempt to guard against cross-side scripting attacks.

**Comment by funderburkjim** [comment]
## Second Fix

There are almost surely circumstances, in addition to those that the first fix addresses, that could occur in the advanced search where all or too many matches might occur.

The safest way to deal with this is simply to omit the 'all' option for the number of returned results.
So, that is what the second fix does.   Now, the maximum number of records returned is 1000.
Surely, for almost all practical purposes 1000 records is ample.

**Comment by funderburkjim** [comment]
## Installed for all dictionaries

As usual, the changes mentioned above were first made in local copy of csl-websanlexicon repository.
Then tested on local server.
Finally, the repository was pushed to Github, pulled to sanskrit-lexicon server.
Then one dictionary was tested on Cologne server, and when all looked well,
the changes were installed for all dictionaries.

**The bug should be fixed now.**

**Comment by funderburkjim** [comment]
Here's a screen shot after the fix (using ap90 dictionary this time).

![image](https://user-images.githubusercontent.com/6393033/65365561-5e26af80-dbe8-11e9-8ed8-b181af7b6642.png)

**Comment by gasyoun** [comment]
> Now, the maximum number of records returned is 1000.

Bad news. So now we will not even know how much entries are found total.

**Comment by funderburkjim** [comment]
You  could do `next` to get the next 1000.

**Comment by gasyoun** [comment]
> next to get the next 1000.

That is pain. And I can't get stats at once.

**Comment by funderburkjim** [comment]
> can't get stats at once

True.  Why don't you open an issue describing this as enhancement.

While we don't want to return all dictionary entries,   it might be possible to return the
total number of matches as a separate, new statistic, independent of the records returned.

**Comment by funderburkjim** [comment]
What are some examples of searches you try that have more than 1000 matches?

**Comment by gasyoun** [comment]
> try that have more than 1000 matches?

`ja` as suffix.

![ja](https://user-images.githubusercontent.com/80761/65794410-c0dff400-e170-11e9-97a7-4e3262658c03.PNG)

**Comment by gasyoun** [comment]
Re-run succeeded; this is not a Tier A source/data correction.

Current source check: advanced-search UI still exposes bounded result sizes up to 1000 in `webtc2/index.php`, and `webtc2/help.html` documents that the Maximum value limits returned matches and that continuation searches may be needed. The thread’s remaining request is for better handling/reporting of very large result sets (for example, total-count statistics for >1000 matches), which is a webtc2 search/display enhancement rather than a dictionary XML/source bug.

Classification: Tier D / enhancement-design. No agent source-correction PR is safe from this issue alone.

**Comment by gasyoun** [comment]
**H1523 hygiene: appears already fixed upstream.** Issue body records Jim's two advanced-search changes (example query fix + CPU/result limit). Residual enhancement (total-count UX) is separate. Closing as completed; reopen if the CPU bomb still reproduces on current webtc2.
